Serialization: Move upgrade test logic to a function

lgoettgens opened the pull request:
This allows to test the upgrading of experimental types as part of the respective experimental folder tests; thus allowing to revert the changes to `experimental/NoExperimental_whitelist_.jl` from https://github.com/oscar-system/Oscar.jl/pull/4162 and thereby re-enabling the `NoExperimental` CI-tests for `LieAlgebras` and `QuadFormAndIso`.

Includes and thus closes #4727.

lgoettgens committed:
Sort `Artifacts.toml` alphabetically

lgoettgens committed:
Move 1.4.0 upgrade tests to appropriate place

lgoettgens committed:
Move upgrade test logic to a function

lgoettgens committed:
Move upgrade tests for experimental types to the respective folders

antonydellavecchia commented on experimental/LieAlgebras/test/Serialization-upgrade-test.jl:
```diff
@@ -0,0 +1,16 @@
+    test_1_4_0_upgrade(;
+      only=[
+        "AbstractLieAlgebra",
+        "AbstractLieAlgebra",
```
```suggestion
        "AbstractLieAlgebraElem",
```

lgoettgens replied:
Oops, good catch!

antonydellavecchia committed:
Update experimental/LieAlgebras/test/Serialization-upgrade-test.jl

antonydellavecchia commented on test/Serialization/upgrades/runtests.jl:
```diff
@@ -34,4 +34,16 @@
+  @testset "< 1.4.0 Upgrade" begin
+    test_1_4_0_upgrade(; exclude=[
+      # upgrading the following is tested in experimental/LieAlgebras/test/Serialization-upgrade-test.jl
+      "AbstractLieAlgebra", "AbstractLieAlgebra",
```
```suggestion
      "AbstractLieAlgebra", "AbstractLieAlgebraElem",
```

antonydellavecchia reviewed (approved):
One last typo, but looks good

lgoettgens committed:
Update test/Serialization/upgrades/runtests.jl

Co-authored-by: antonydellavecchia <antonydellavecchia@gmail.com>

lgoettgens commented:
There are still issues with the Vector tests, since vectors may have elems from the LieAlgebra folder.

Some ideas come to my mind, but some require changes to the tarball:
1. Split the folder `Vector` into multiple ones based on the eltype.
2. Split the whole tarball into three folders (src, LieAlgebras, QuadFormAndIsom) each with the current structure. This would simplify the whole function added here.
3. Make the `only` and `exclude` kwargs added here not only take type strings, but also file IDs inside the folders.

From the perspective of things gradually moving from experimental to src in the future, I wouldn't be very happy with option 2, as that may require changes to the tarball in the future again.

I'll try to update this here according to 3 tomorrow.

lgoettgens committed:
Fix Vector upgrade tests

lgoettgens committed:
Reset serializer state in the end

lgoettgens commented:
This is now adapted according to proposal 3.

Furthermore, I included https://github.com/oscar-system/Oscar.jl/pull/4727 here, since merging that would otherwise lead to a conflict.
